ikura/shim/src/dock/sovereign.rs

```diff
@@ -81,14 +89,36 @@ impl SovereignRPCServer for SovereignDock {
             .as_ref()
             .cloned()
             .ok_or_else(err::no_signing_key)?;
+        let nonce = self.gen_nonce().await.map_err(err::nonce_obtain_error)?;
+        let blob_extrinsic = self
+            .client
+            .make_blob_extrinsic(blob, namespace, &submit_key, nonce)
+            .await
+            .map_err(err::submit_extrinsic_error)?;
         self.client
-            .submit_blob(blob, namespace, submit_key)
+            .submit_blob(&blob_extrinsic)
             .await
             .map_err(err::submission_error)?;
         Ok(())
     }
 }
 
+impl SovereignDock {
+    async fn gen_nonce(&self) -> anyhow::Result<u64> {
+        let submit_key = self
+            .submit_key
+            .as_ref()
+            .ok_or_else(|| anyhow::anyhow!("no key for signing blobs"))?; // should be unreachable
+        let mut cur_nonce = self.cur_nonce.lock().await;
+        let nonce = match *cur_nonce {
+            Some(nonce) => nonce,
+            None => self.client.get_last_nonce(&submit_key).await?,
+        };
+        cur_nonce.replace(nonce + 1);
```

Comment:
After the first submitted blob, the nonce is increased by one automatically, but this could lead to breaking the shim if the same account used in the shim to sign a transaction is used for any other purposes within the chain outside of the shim.

I'm not entirely sure, but this could potentially happen with teleportation. Sending tokens from Kusama to Ikura should not be a problem because the nonce would only increase within Kusama, and the tokens are deposited into the Ikura chain. However, the opposite scenario could break the shim. If an account used in the shim wants to send funds back to Kusama, it needs to shut down the shim, perform the operation, and then restart it. Otherwise, the increased nonce that occurred in the XCM transaction would not be recognized by the shim

Reply:
Oh, you described this problem in the following PR!

Reply:
Well, nonce handling requires exclusive control over the account. It's not a problem of this particular piece of code, it's the general problem. If you launch two instances of, e.g. `query submit`, with almost 100% certainty you can be sure that one of those will fail.

You can try to avoid that by trying to share the nonce somewhere e.g. under `$HOME`. While that may work, even if tricky to implement, it won't work for other cases: e.g. if the user tried to run the command under another user or machine. You may think that you can look into the mempool and try to figure out the latest nonce from that, but IMO that's futile, and won't ever work. I don't see a fault proof way to avoid that.

Instead, you should be looking at the landed blobs and re-sign them. This is however a bit too much for this PR and there is an entire catch-all issue regarding concerns like that https://github.com/thrumdev/blobs/issues/79

Pragmatic solution for this right now IMO is to try to give heads up to the user, which I tried to do in https://github.com/thrumdev/blobs/pull/271